File: RoboticsExercises/Exercise3/Solution/reachabilityAnalysis.py
import sdurw
import sdurw_proximitystrategies
import numpy as np

def getConfigurations(nameGoal, nameTcp, robot, wc, state):
    # Get, make and print name of frames
    robotName = robot.getName()
    nameRobotBase = robotName + "." + "Base"
    nameRobotTcp = robotName + "." + "TCP"

    # Find frames and check for existence
    frameGoal = wc.findFrame(nameGoal)
    frameTcp = wc.findFrame(nameTcp)
    frameRobotBase = wc.findFrame(nameRobotBase)
    frameRobotTcp = wc.findFrame(nameRobotTcp)
    if frameGoal == None or frameTcp == None or frameRobotBase == None or frameRobotTcp == None:
        print(" ALL FRAMES NOT FOUND:")
        print(" Found \"", nameGoal, "\": ", "NO!" if frameGoal == None else "YES!")
        print(" Found \"", nameTcp, "\": ", "NO!" if frameTcp == None else "YES!")
        print(" Found \"", nameRobotBase, "\": ", "NO!" if frameRobotBase == None else "YES!")
        print(" Found \"", nameRobotTcp, "\": ", "NO!" if frameRobotTcp == None else "YES!")

    # Make "helper" transformations
    frameBaseTGoal = sdurw.Kinematics.frameTframe(frameRobotBase, frameGoal, state)
    frameTcpTRobotTcp = sdurw.Kinematics.frameTframe(frameTcp, frameRobotTcp, state)

    # get grasp frame in robot tool frame
    targetAt = frameBaseTGoal * frameTcpTRobotTcp
    closedFormSovler = sdurw.ClosedFormIKSolverUR(robot, state)
    return closedFormSovler.solve(targetAt, state)



if __name__ == "__main__":
    #load workcell
    wc = sdurw.WorkCellLoaderFactory.load("./scene/Scene.wc.xml")
    if wc.isNull():
        raise Exception("COULD NOT LOAD scene... check path!")

    # find relevant frames
    cylinderFrame = wc.findMovableFrame("Cylinder")
    if cylinderFrame == None:
        raise Exception("COULD not find movable frame Cylinder ... check model")

    robotUR5 = wc.findSerialDevice("UR5")
    if robotUR5.isNull():
        raise Exception("COULD not find device UR5 ... check model")

    detector = sdurw.CollisionDetector(wc, sdurw_proximitystrategies.ProximityStrategyFactory_makeDefaultCollisionStrategy())


    # get the default state
    state = wc.getDefaultState()
    collisionFreeSolutions = []

    for rollAngle in range(360): # for every degree around the roll axis

        cylinderFrame.setTransform(sdurw.Transform3D(cylinderFrame.getTransform(state).P(), sdurw.RPY(np.deg2rad(rollAngle),0,0)), state)


        solutions = getConfigurations("GraspTarget", "GraspTCP", robotUR5, wc, state)

        for i in range(len(solutions)):
            # set the robot in that configuration and check if it is in collision
            robotUR5.setQ(solutions[i], state)
            res1 = sdurw.ProximityData()
            if not detector.inCollision(state, res1):
                collisionFreeSolutions.append(solutions[i]) # save it
                break # we only need one


    print("Current position of the robot vs object to be grasped has: ", len(collisionFreeSolutions), " collision-free inverse kinematics solutions!")


    # visualize them
    tStatePath = sdurw.PathTimedState()
    time=0
    for i in range(len(collisionFreeSolutions)):
        robotUR5.setQ(collisionFreeSolutions[i], state)
        tStatePath.push_back(sdurw.TimedState(time,state))
        time+=0.01

    sdurw.PathLoader.storeTimedStatePath(wc.deref(),tStatePath,"./scene/visu.rwplay")

File: RoboticsExercises/Exercise3/Solution/reachabilityAnalysis.cpp

#include <rw/invkin.hpp>
#include <rw/rw.hpp>
#include <rwlibs/proximitystrategies/ProximityStrategyFactory.hpp>

#include <iostream>
#include <string>
#include <filesystem>

using namespace rw::kinematics;
using namespace rw::math;

std::vector< Q > getConfigurations (const std::string nameGoal, const std::string nameTcp,
                                    rw::models::SerialDevice::Ptr robot,
                                    rw::models::WorkCell::Ptr wc, State& state)
{
    // Get, make and print name of frames
    const std::string robotName     = robot->getName ();
    const std::string nameRobotBase = robotName + "." + "Base";
    const std::string nameRobotTcp  = robotName + "." + "TCP";

    // Find frames and check for existence
    Frame::Ptr goal_f      = wc->findFrame (nameGoal);
    Frame::Ptr tcp_f       = wc->findFrame (nameTcp);
    Frame::Ptr robotBase_f = wc->findFrame (nameRobotBase);
    Frame::Ptr robotTcp_f  = wc->findFrame (nameRobotTcp);
    if (goal_f.isNull () || tcp_f.isNull () || robotBase_f.isNull () || robotTcp_f.isNull ()) {
        std::cout << " ALL FRAMES NOT FOUND:" << std::endl;
        std::cout << " Found \"" << nameGoal << "\": " << (goal_f.isNull () ? "NO!" : "YES!")
                  << std::endl;
        std::cout << " Found \"" << nameTcp << "\": " << (tcp_f.isNull () ? "NO!" : "YES!")
                  << std::endl;
        std::cout << " Found \"" << nameRobotBase
                  << "\": " << (robotBase_f.isNull () ? "NO!" : "YES!") << std::endl;
        std::cout << " Found \"" << nameRobotTcp
                  << "\": " << (robotTcp_f.isNull () ? "NO!" : "YES!") << std::endl;
    }

    // Make "helper" transformations
    Transform3D<> baseTGoal    = Kinematics::frameTframe (robotBase_f, goal_f, state);
    Transform3D<> tcpTRobotTcp = Kinematics::frameTframe (tcp_f, robotTcp_f, state);

    // get grasp frame in robot tool frame
    Transform3D<> targetAt = baseTGoal * tcpTRobotTcp;

    rw::invkin::ClosedFormIKSolverUR::Ptr closedFormSovler =
        rw::core::ownedPtr (new rw::invkin::ClosedFormIKSolverUR (robot, state));

    return closedFormSovler->solve (targetAt, state);
}

int main (int argc, char** argv)
{
    // load workcell
    rw::models::WorkCell::Ptr wc =
        rw::loaders::WorkCellLoader::Factory::load (std::filesystem::path(__FILE__).parent_path()/"scene/Scene.wc.xml");
    if (wc.isNull ()) {
        RW_THROW ("COULD NOT LOAD scene... check path!");
    }

    // find relevant frames
    MovableFrame::Ptr cylinderFrame = wc->findFrame< MovableFrame > ("Cylinder");
    if (cylinderFrame.isNull ()) {
        RW_THROW ("COULD not find movable frame Cylinder ... check model");
    }

    rw::models::SerialDevice::Ptr robotUR5 = wc->findDevice< rw::models::SerialDevice > ("UR5");
    if (robotUR5.isNull ()) {
        RW_THROW ("COULD not find device UR5 ... check model");
    }

    // create Collision Detector
    rw::proximity::CollisionDetector detector (
        wc, rwlibs::proximitystrategies::ProximityStrategyFactory::makeDefaultCollisionStrategy ());

    // get the default state
    State state = wc->getDefaultState ();
    std::vector< Q > collisionFreeSolutions;

    for (double rollAngle = 0; rollAngle < 360.0;
         rollAngle += 1.0) {    // for every degree around the roll axis

        cylinderFrame->setTransform (Transform3D<> (Vector3D<> (cylinderFrame->getTransform (state).P ()),
                                              RPY<> (rollAngle * Deg2Rad, 0, 0)),
                               state);

        std::vector< Q > solutions =
            getConfigurations ("GraspTarget", "GraspTCP", robotUR5, wc, state);

        for (unsigned int i = 0; i < solutions.size (); i++) {
            // set the robot in that configuration and check if it is in collision
            robotUR5->setQ (solutions[i], state);

            if (!detector.inCollision (state)) {
                collisionFreeSolutions.push_back (solutions[i]);    // save it
                break;                                              // we only need one
            }
        }
    }

    std::cout << "Current position of the robot vs object to be grasped has: "
              << collisionFreeSolutions.size () << " collision-free inverse kinematics solutions!"
              << std::endl;

    // visualize them
    rw::trajectory::TimedStatePath tStatePath;
    double time = 0;
    for (unsigned int i = 0; i < collisionFreeSolutions.size (); i++) {
        robotUR5->setQ (collisionFreeSolutions[i], state);
        tStatePath.push_back (rw::trajectory::TimedState (time, state));
        time += 0.01;
    }

    rw::loaders::PathLoader::storeTimedStatePath (*wc, tStatePath, "./visu.rwplay");

    return 0;
}
